docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -136,11 +140,38 @@ for given scenarios are elaborated in more detail below.
 
 TODO: Async iterators (returning `IAsyncEnumerable<T>`)
 
-#### `Task`, `Task<T>`, `ValueTask`, `ValueTask<T>` Scenarios
-
-For any lvalue of one of these types, we'll generally rewrite `await expr` into `System.Runtime.CompilerServices.RuntimeHelpers.Await(expr)`. A number of different example scenarios for this are covered below. The
+#### `RuntimeHelpers.Await` Scenarios
+
+For any `await expr` with where `expr` has type `E`, the compiler will attempt to match it to a helper method in `System.Runtime.CompilerServices.RuntimeHelpers`. The following algorithm is used:
+
+1. If `E` has generic arity greater than 1, no match is found and instead move to [await any other type].
+2. `System.Runtime.CompilerServices.RuntimeHelpers` from corelib (the library that defines `System.Object` and has no references) is fetched.
+3. All methods named `Await` are put into a group called `M`.
+4. For every `Mi` in `M`:
+   1. If `Mi`'s generic arity does not match `E`, it is removed.
+   2. If `Mi` takes more than 1 parameter (named `P`), it is removed.
+   3. If `Mi` has a generic arity of 0, all of the following must be true, or `Mi` is removed:
+      1. The return type is `System.Void`
+      2. There is an identity or implicit reference conversion from `E` to the type of `P`.
+   4. Otherwise, if `Mi` has a generic arity of 1 with type param `Tm`, all of the following must be true, or `Mi` is removed:
+      1. The return type is `Tm`
+      2. There is an identity or implicit reference conversion from `E`'s unsubstituted definition to `P`
```

Comment:
I'm unsure what "`E`'s unsubstituted definition" means. Is it something like, if `E` is `Task<C>` then its unsubstituted definition is `Task<T>`?

Reply:
Correct. I also struggled to find a better wording, but couldn't come up with one.